### ISSUE TITLE
filter_samples in phyloseq

### ISSUE THREAD

**Issue by rozlynboutin** [joey711/phyloseq#1167]
Hi there,

I have a very basic question and I apologize for redundancy if this has been previously addressed, but I couldn't find an issue with quite the same problem. I am trying to filter my phyloseq object to include features with a total relative abundance >1e-5 using the following code: 
physeqITSr  = transform_sample_counts(physeqITS_DADA, function(x) x / sum(x))
physeqITSfr = filter_taxa(physeqITSr, function(x) mean(x) > 1e-5, TRUE)

I am getting the error:

Error in validObject(.Object) : invalid class “otu_table” object: 
 OTU abundance data must have non-zero dimensions.

If I export my physeqITSr object as a data frame, however, everything looks the way it should. I do have a few samples that have Na for the features though since they were removed from the analysis in previous steps-could this be the source of the problem?

Thanks in advance!

**Comment by mikemc** [comment]
The error
> Error in validObject(.Object) : invalid class “otu_table” object:
> OTU abundance data must have non-zero dimensions.
indicates that OTUs are passing the condition supplied to `filter_taxa()`. It is possible that the NAs are causing a problem, since I don't think phyloseq is designed to have NAs in the otu_table's. To check, you might try
```r
passing_otus <- filter_taxa(physeqITSr, function(x) mean(x) > 1e-5, prune = FALSE)
```
I would guess that any OTU that has an NA in any sample will show up as NA in this list and be filtered when you use the `prune=TRUE` option. So, if you have one sample with NA for everything, that will result in all taxa being filtered.

The specific filter criterion you are using can also result in 0 or very few OTUs passing. Microbiome data is often sparse, meaning that most taxa will have an abundance of 0 in most samples, and thus overall very low average proportion. Rather than requiring `function(x) mean(x) > 1e-5` I would typically use something like `function(x) max(x) > 1e-3`, perhaps in conjunction with a prevalence requirement e.g. `function(x) sum(x>1e-3) > 5`. This way you keep a taxa that are at a high enough frequency in at least one or a couple samples, even if it is very rare across all samples.

**Comment by rozlynboutin** [comment]
Hi!

Thanks very much for the quick reply and for the suggestions! I think I'd like to first fix the NA issue and then I can play with the filtering. From what I can tell, the NAs are being introduced when I convert my table to relative abundances. By converting my phyloseq otu table to a data frame at each step, I've determined that I seem to lose a lot of reads for certain samples (which later become NA) when I import my biom file into phyloseq. I created my biome file in qiime2 and export it as both a regular biom file and as a json biom file. If I convert these to text files, my read numbers for each feature are normal, but once I import these into phyloseq, the read numbers are being changed. Do you know what might be causing this?

My code is below:
biom_file<- "table--ITS2-full_json.biom"
biom_otu_tax <- import_biom(biom_file, parseFunction = parse_taxonomy_default)


Thanks in advance!

**Comment by mikemc** [comment]
I strongly suspect that there are NAs before you call `transform_sample_counts()`. Otherwise, filtering the NAs first wouldn't make a difference. You might be not seeing them before converting to proportions: If there is just one NA for one taxon for one sample, then all the taxa for that sample will become NA after you try to normalize that sample to proportions. Try running the following to see if there are any NAs,
```r
sum(is.na(otu_table(physeqITS_DADA)))
```

**Comment by rozlynboutin** [comment]
Thanks again for your help! I've run the code you suggested to check for NAs before normalizing to proportions and it is equal to 0, even when I run everything from the point of loading the data as a phyloseq object. So, there are no NAs before I run 

physeqITSr = transform_sample_counts(physeqITS_DADA, function(x) x / sum(x))

Would an NA end up in the table for a sample after this step if that sample had only one OTU assigned only at the Kingdom level? (I do this taxonomy filtering step later)

**Comment by mikemc** [comment]
I'm pretty sure the `tax_table` should not affect the call to `transform_sample_counts()`. The only other option I can think of is that those samples have total abundances of 0, which you can check with `sample_sums()`. But that should be pretty obvious and you should be ending up with `NaN`s instead of `NA`s. If that's not what's happening and there really are no NAs or NaNs in the table then if this sounds like a bug, and I would need to be able to reproduce it myself to figure out what is going on. If you link to an .Rds file of `physeqITS_DADA` (made with `saveRDS(physeqITS_DADA, "physeqITS_DADA.Rds")` then I can take a look. Also, please post the phyloseq version `packageVersion("phyloseq")`.

**Comment by rozlynboutin** [comment]
You are right! The sample_sums() for 5 of my samples was 0 and then 2 other samples were affected after I filtered out taxa present in only 1 sample-thanks so much! If I filter these samples out of the table first, everything works as it should :) 

I have one other small question-I am looking at fungal ITS2 sequencing data and imported my biom file from qiime2 after appending the ASV taxonomy assigned from the UNITE database to the table.  My rank_names for the taxonomy come up as "Rank1", "Rank2", etc. instead of "Kingdom", "Phylum", etc when I import the table using:

biom_otu_tax_DADA <- import_biom(biom_file, parseFunction = parse_taxonomy_default)

Is there a way to change the annotation? 

Thanks in advance!

**Comment by mikemc** [comment]
It may be that if you use the `parse_taxonomy_qiime` function instead of `parse_taxonomy_default` that will solve the problem. Try that, or looking at the help for the import_biom and parse_taxonomy functions. The info on importing data in the tutorials and vignettes (`browseVignettes("phyloseq")` might also help.

**Comment by rozlynboutin** [comment]
Ok, thanks so much! I played around with the function and it works now!
